From edce05d7566cc93c90ddd37013fcf4e99d75eb62 Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Thu, 4 Nov 2021 21:13:10 +0500
Subject: [PATCH] Add logging on startup


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 71c982ba..d5480cb5 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -152,6 +152,8 @@ public class BungeeCord extends ProxyServer
         registerChannel( "BungeeCord" );
     }
 
+    private final long serverStartTime;
+
     public static BungeeCord getInstance()
     {
         return (BungeeCord) ProxyServer.getInstance();
@@ -200,6 +202,9 @@ public class BungeeCord extends ProxyServer
         logger = io.github.waterfallmc.waterfall.log4j.WaterfallLogger.create();
         // Waterfall end
 
+        logger.info("Starting " + getName() + " version " + getVersion() + " (MC:1.7.x - 1.18.x)");
+        serverStartTime = System.nanoTime();
+
         pluginManager = new PluginManager( this );
         // FlameCord - We register commands in our new method
         //getPluginManager().registerCommand( null, new CommandReload() );
@@ -304,6 +309,8 @@ public class BungeeCord extends ProxyServer
                 independentThreadStop( getTranslation( "info_restart" ), false );
             }
         } );
+        String doneTime = String.format(Locale.ROOT, "%.3fs", (double) (System.nanoTime() - serverStartTime) / 1.0E9D);
+        logger.info("Done (" + doneTime + ")!");
     }
 
     public void startListeners()
-- 
2.33.0

