From 83fdfb3c75706f914e3761476650165569e284a8 Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Sun, 2 Jan 2022 16:20:42 +0500
Subject: [PATCH] Finish early to avoid multiple incoming packets


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 3c7adbcf..4ba8111d 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -439,6 +439,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
     public void handle(final EncryptionResponse encryptResponse) throws Exception
     {
         Preconditions.checkState( thisState == State.ENCRYPT, "Not expecting ENCRYPT" );
+        thisState = State.FINISHING; // Amethyst - Finish here to avoid multiple incoming packets
 
         SecretKey sharedKey = EncryptionUtil.getSecret( encryptResponse, request );
         // Waterfall start
@@ -499,6 +500,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
 
     private void finish()
     {
+        thisState = State.FINISHING; // Amethyst - Finish here to avoid multiple incoming packets
         if ( isOnlineMode() )
         {
             // Check for multiple connections
-- 
2.34.1.windows.1

