From 7380f80196877ab46095bc34fc109fc1c83c03b6 Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Sun, 2 Jan 2022 16:22:55 +0500
Subject: [PATCH] Change IllegalStateException to QuietException


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index b44d13e5..bd142afa 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -66,6 +66,7 @@ import net.md_5.bungee.tab.ServerUnique;
 import net.md_5.bungee.tab.TabList;
 import net.md_5.bungee.util.CaseInsensitiveSet;
 import net.md_5.bungee.util.ChatComponentTransformer;
+import net.md_5.bungee.util.QuietException;
 
 @RequiredArgsConstructor
 public final class UserConnection implements ProxiedPlayer
@@ -311,7 +312,9 @@ public final class UserConnection implements ProxiedPlayer
 
             if ( getServer() == null && !ch.isClosing() )
             {
-                throw new IllegalStateException( "Cancelled ServerConnectEvent with no server or disconnect." );
+                // Amethyst start - Change IllegalStateException to QuietException and explain that is a plugin
+               // throw new IllegalStateException( "Cancelled ServerConnectEvent with no server or disconnect." );
+                throw new QuietException( "A plugin cancelled ServerConnectEvent with no server or disconnect." );
             }
             return;
         }
-- 
2.36.1.windows.1

