From 703cfcbe11ccff5a3d180e44101b5e65b67c32fa Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Sun, 2 Jan 2022 16:26:25 +0500
Subject: [PATCH] Don't declare uuid unsess it's null


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index c8a5c651..32515b0f 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -534,7 +534,13 @@ public class InitialHandler extends PacketHandler implements PendingConnection
                     {
                         loginProfile = obj;
                         name = obj.getName();
-                        uniqueId = Util.getUUID( obj.getId() );
+                        // Amethyst start - Don't declare uuid unless it's null
+                       // uniqueId = Util.getUUID( obj.getId() );
+                        if (uniqueId == null)
+                        {
+                            uniqueId = Util.getUUID( obj.getId() );
+                        }
+                        // Amethyst end
                         finish();
                         return;
                     }
-- 
2.37.1.windows.1

