From 9da43986d76d9391b2a06d2936523805b86dc5b7 Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Sun, 2 Jan 2022 16:26:25 +0500
Subject: [PATCH] Don't declare uuid unsess it's null


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 3c7adbcf..316a50a0 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -481,7 +481,13 @@ public class InitialHandler extends PacketHandler implements PendingConnection
                     {
                         loginProfile = obj;
                         name = obj.getName();
-                        uniqueId = Util.getUUID( obj.getId() );
+                        // Amethyst start - Don't declare uuid unless it's null
+                       // uniqueId = Util.getUUID( obj.getId() );
+                        if (uniqueId == null)
+                        {
+                            uniqueId = Util.getUUID( obj.getId() );
+                        }
+                        // Amethyst end
                         finish();
                         return;
                     }
-- 
2.35.1

