From a6028039e0bc8e367cc479d8fda8bc41eeef46c1 Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Tue, 4 Jan 2022 20:41:35 +0500
Subject: [PATCH] Amethyst modules system


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index e7567771..c6bbbe7e 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -28,9 +28,55 @@ import net.md_5.bungee.config.ConfigurationProvider;
  * @author DenaryDev
  */
 public class AmethystConfig extends AbstractYamlConfig {
+    /**
+     * A module that provides the /alert and /alertraw commands.
+     */
+    public boolean alertEnabled = true;
+    /**
+     * A module that provides the /find command.
+     */
+    public boolean findEnabled = true;
+    /**
+     * A module that provides the /ip command.
+     */
+    public boolean ipEnabled = true;
+    /**
+     * A module that provides the /glist command.
+     */
+    public boolean listEnabled = true;
+    /**
+     * A module that provides the /gperms command.
+     */
+    public boolean permsEnabled = true;
+    /**
+     * A module that provides the /send command.
+     */
+    public boolean sendEnabled = true;
+    /**
+     * A module that provides the /server command.
+     */
+    public boolean serverEnabled = true;
+    /**
+     * A module that contains reconnect location functionality in locations.yml.
+     */
+    public boolean reconnectEnabled = false;
 
     public AmethystConfig(ConfigurationProvider provider) {
         super(provider, "amethyst.yml");
         init();
+        loadContent();
+    }
+
+    @Override
+    protected void loadContent()
+    {
+        alertEnabled = getBoolean( "modules.alert", alertEnabled );
+        findEnabled = getBoolean( "modules.find", findEnabled );
+        ipEnabled = getBoolean( "modules.ip", ipEnabled );
+        listEnabled = getBoolean( "modules.list", listEnabled );
+        permsEnabled = getBoolean( "modules.perms", permsEnabled );
+        sendEnabled = getBoolean( "modules.send", sendEnabled );
+        serverEnabled = getBoolean( "modules.server", serverEnabled );
+        reconnectEnabled = getBoolean( "modules.reconnect", reconnectEnabled );
     }
 }
diff --git a/flame/src/main/resources/amethyst.yml b/flame/src/main/resources/amethyst.yml
index 72394640..e8da947f 100644
--- a/flame/src/main/resources/amethyst.yml
+++ b/flame/src/main/resources/amethyst.yml
@@ -1,2 +1,44 @@
 # Amethyst configuration file
 # Generated on ${project.version}
+
+# Modules configuration
+# In this section you can enable or disable certain proxy modules
+modules:
+  # Alert commands module
+  # Provides /alert and /alertraw commands
+  alert: true
+
+  # Find command module
+  # Provides /find command
+  # Default: true
+  find: true
+
+  # Ip command module
+  # Provides /ip command
+  # Default: true
+  ip: true
+
+  # List command module
+  # Provides /glist command
+  # Default: true
+  list: true
+
+  # Perms command module
+  # Provides /gperms command
+  # Default: true
+  perms: true
+
+  # Send command module
+  # Provides /send command
+  # Default: true
+  send: true
+
+  # Server command module
+  # Provides /server command
+  # Default: true
+  server: true
+
+  # Reconnect yaml module
+  # Provides reconnect location functionality in locations.yml
+  # Default: false
+  reconnect: false
diff --git a/module/cmd-alert/pom.xml b/module/cmd-alert/pom.xml
deleted file mode 100644
index c0d9f7d5..00000000
--- a/module/cmd-alert/pom.xml
+++ /dev/null
@@ -1,19 +0,0 @@
-
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
-    <modelVersion>4.0.0</modelVersion>
-
-    <parent>
-        <groupId>io.sapphiremc.amethyst</groupId>
-        <artifactId>amethyst-module</artifactId>
-        <version>1.18-R0.1-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
-    </parent>
-
-    <artifactId>amethyst-module-cmd-alert</artifactId>
-    <version>1.18-R0.1-SNAPSHOT</version>
-    <packaging>jar</packaging>
-
-    <name>cmd_alert</name>
-    <description>Provides the alert and alertraw commands</description>
-</project>
diff --git a/module/cmd-alert/src/main/java/net/md_5/bungee/module/cmd/alert/PluginAlert.java b/module/cmd-alert/src/main/java/net/md_5/bungee/module/cmd/alert/PluginAlert.java
deleted file mode 100644
index b6f84865..00000000
--- a/module/cmd-alert/src/main/java/net/md_5/bungee/module/cmd/alert/PluginAlert.java
+++ /dev/null
@@ -1,14 +0,0 @@
-package net.md_5.bungee.module.cmd.alert;
-
-import net.md_5.bungee.api.plugin.Plugin;
-
-public class PluginAlert extends Plugin
-{
-
-    @Override
-    public void onEnable()
-    {
-        getProxy().getPluginManager().registerCommand( this, new CommandAlert() );
-        getProxy().getPluginManager().registerCommand( this, new CommandAlertRaw() );
-    }
-}
diff --git a/module/cmd-alert/src/main/resources/bungee.yml b/module/cmd-alert/src/main/resources/bungee.yml
deleted file mode 100644
index 9dcf1704..00000000
--- a/module/cmd-alert/src/main/resources/bungee.yml
+++ /dev/null
@@ -1,5 +0,0 @@
-name: ${project.name}
-main: net.md_5.bungee.module.cmd.alert.PluginAlert
-version: ${project.version}
-description: ${project.description}
-author: ${module.author}
diff --git a/module/cmd-find/pom.xml b/module/cmd-find/pom.xml
deleted file mode 100644
index e8172a66..00000000
--- a/module/cmd-find/pom.xml
+++ /dev/null
@@ -1,19 +0,0 @@
-
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
-    <modelVersion>4.0.0</modelVersion>
-
-    <parent>
-        <groupId>io.sapphiremc.amethyst</groupId>
-        <artifactId>amethyst-module</artifactId>
-        <version>1.18-R0.1-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
-    </parent>
-
-    <artifactId>amethyst-module-cmd-find</artifactId>
-    <version>1.18-R0.1-SNAPSHOT</version>
-    <packaging>jar</packaging>
-
-    <name>cmd_find</name>
-    <description>Provides the find command</description>
-</project>
diff --git a/module/cmd-find/src/main/java/net/md_5/bungee/module/cmd/find/PluginFind.java b/module/cmd-find/src/main/java/net/md_5/bungee/module/cmd/find/PluginFind.java
deleted file mode 100644
index 63757214..00000000
--- a/module/cmd-find/src/main/java/net/md_5/bungee/module/cmd/find/PluginFind.java
+++ /dev/null
@@ -1,13 +0,0 @@
-package net.md_5.bungee.module.cmd.find;
-
-import net.md_5.bungee.api.plugin.Plugin;
-
-public class PluginFind extends Plugin
-{
-
-    @Override
-    public void onEnable()
-    {
-        getProxy().getPluginManager().registerCommand( this, new CommandFind() );
-    }
-}
diff --git a/module/cmd-find/src/main/resources/bungee.yml b/module/cmd-find/src/main/resources/bungee.yml
deleted file mode 100644
index 34efb98d..00000000
--- a/module/cmd-find/src/main/resources/bungee.yml
+++ /dev/null
@@ -1,5 +0,0 @@
-name: ${project.name}
-main: net.md_5.bungee.module.cmd.find.PluginFind
-version: ${project.version}
-description: ${project.description}
-author: ${module.author}
diff --git a/module/cmd-list/pom.xml b/module/cmd-list/pom.xml
deleted file mode 100644
index a222f913..00000000
--- a/module/cmd-list/pom.xml
+++ /dev/null
@@ -1,19 +0,0 @@
-
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
-    <modelVersion>4.0.0</modelVersion>
-
-    <parent>
-        <groupId>io.sapphiremc.amethyst</groupId>
-        <artifactId>amethyst-module</artifactId>
-        <version>1.18-R0.1-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
-    </parent>
-
-    <artifactId>amethyst-module-cmd-list</artifactId>
-    <version>1.18-R0.1-SNAPSHOT</version>
-    <packaging>jar</packaging>
-
-    <name>cmd_list</name>
-    <description>Provides the glist command</description>
-</project>
diff --git a/module/cmd-list/src/main/java/net/md_5/bungee/module/cmd/list/PluginList.java b/module/cmd-list/src/main/java/net/md_5/bungee/module/cmd/list/PluginList.java
deleted file mode 100644
index e2a196c1..00000000
--- a/module/cmd-list/src/main/java/net/md_5/bungee/module/cmd/list/PluginList.java
+++ /dev/null
@@ -1,13 +0,0 @@
-package net.md_5.bungee.module.cmd.list;
-
-import net.md_5.bungee.api.plugin.Plugin;
-
-public class PluginList extends Plugin
-{
-
-    @Override
-    public void onEnable()
-    {
-        getProxy().getPluginManager().registerCommand( this, new CommandList() );
-    }
-}
diff --git a/module/cmd-list/src/main/resources/bungee.yml b/module/cmd-list/src/main/resources/bungee.yml
deleted file mode 100644
index c5999f77..00000000
--- a/module/cmd-list/src/main/resources/bungee.yml
+++ /dev/null
@@ -1,5 +0,0 @@
-name: ${project.name}
-main: net.md_5.bungee.module.cmd.list.PluginList
-version: ${project.version}
-description: ${project.description}
-author: ${module.author}
diff --git a/module/cmd-send/pom.xml b/module/cmd-send/pom.xml
deleted file mode 100644
index 6e864859..00000000
--- a/module/cmd-send/pom.xml
+++ /dev/null
@@ -1,19 +0,0 @@
-
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
-    <modelVersion>4.0.0</modelVersion>
-
-    <parent>
-        <groupId>io.sapphiremc.amethyst</groupId>
-        <artifactId>amethyst-module</artifactId>
-        <version>1.18-R0.1-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
-    </parent>
-
-    <artifactId>amethyst-module-cmd-send</artifactId>
-    <version>1.18-R0.1-SNAPSHOT</version>
-    <packaging>jar</packaging>
-
-    <name>cmd_send</name>
-    <description>Provides the gsend command</description>
-</project>
diff --git a/module/cmd-send/src/main/java/net/md_5/bungee/module/cmd/send/PluginSend.java b/module/cmd-send/src/main/java/net/md_5/bungee/module/cmd/send/PluginSend.java
deleted file mode 100644
index 4d4cf822..00000000
--- a/module/cmd-send/src/main/java/net/md_5/bungee/module/cmd/send/PluginSend.java
+++ /dev/null
@@ -1,13 +0,0 @@
-package net.md_5.bungee.module.cmd.send;
-
-import net.md_5.bungee.api.plugin.Plugin;
-
-public class PluginSend extends Plugin
-{
-
-    @Override
-    public void onEnable()
-    {
-        getProxy().getPluginManager().registerCommand( this, new CommandSend() );
-    }
-}
diff --git a/module/cmd-send/src/main/resources/bungee.yml b/module/cmd-send/src/main/resources/bungee.yml
deleted file mode 100644
index 8f0d0edd..00000000
--- a/module/cmd-send/src/main/resources/bungee.yml
+++ /dev/null
@@ -1,5 +0,0 @@
-name: ${project.name}
-main: net.md_5.bungee.module.cmd.send.PluginSend
-version: ${project.version}
-description: ${project.description}
-author: ${module.author}
diff --git a/module/cmd-server/pom.xml b/module/cmd-server/pom.xml
deleted file mode 100644
index b0d3bff1..00000000
--- a/module/cmd-server/pom.xml
+++ /dev/null
@@ -1,19 +0,0 @@
-
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
-    <modelVersion>4.0.0</modelVersion>
-
-    <parent>
-        <groupId>io.sapphiremc.amethyst</groupId>
-        <artifactId>amethyst-module</artifactId>
-        <version>1.18-R0.1-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
-    </parent>
-
-    <artifactId>amethyst-module-cmd-server</artifactId>
-    <version>1.18-R0.1-SNAPSHOT</version>
-    <packaging>jar</packaging>
-
-    <name>cmd_server</name>
-    <description>Provides the server command</description>
-</project>
diff --git a/module/cmd-server/src/main/java/net/md_5/bungee/module/cmd/server/PluginServer.java b/module/cmd-server/src/main/java/net/md_5/bungee/module/cmd/server/PluginServer.java
deleted file mode 100644
index 4aaae7ec..00000000
--- a/module/cmd-server/src/main/java/net/md_5/bungee/module/cmd/server/PluginServer.java
+++ /dev/null
@@ -1,13 +0,0 @@
-package net.md_5.bungee.module.cmd.server;
-
-import net.md_5.bungee.api.plugin.Plugin;
-
-public class PluginServer extends Plugin
-{
-
-    @Override
-    public void onEnable()
-    {
-        getProxy().getPluginManager().registerCommand( this, new CommandServer() );
-    }
-}
diff --git a/module/cmd-server/src/main/resources/bungee.yml b/module/cmd-server/src/main/resources/bungee.yml
deleted file mode 100644
index b0b87841..00000000
--- a/module/cmd-server/src/main/resources/bungee.yml
+++ /dev/null
@@ -1,5 +0,0 @@
-name: ${project.name}
-main: net.md_5.bungee.module.cmd.server.PluginServer
-version: ${project.version}
-description: ${project.description}
-author: ${module.author}
diff --git a/module/pom.xml b/module/pom.xml
deleted file mode 100644
index 9117dc2b..00000000
--- a/module/pom.xml
+++ /dev/null
@@ -1,53 +0,0 @@
-
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
-    <modelVersion>4.0.0</modelVersion>
-
-    <parent>
-        <groupId>io.sapphiremc.amethyst</groupId>
-        <artifactId>amethyst-parent</artifactId>
-        <version>1.18-R0.1-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
-    </parent>
-
-    <artifactId>amethyst-module</artifactId>
-    <version>1.18-R0.1-SNAPSHOT</version>
-    <packaging>pom</packaging>
-
-    <name>Amethyst Modules</name>
-    <description>Parent project for all Amethyst modules.</description>
-
-    <modules>
-        <module>cmd-alert</module>
-        <module>cmd-find</module>
-        <module>cmd-list</module>
-        <module>cmd-send</module>
-        <module>cmd-server</module>
-        <module>reconnect-yaml</module>
-    </modules>
-
-    <properties>
-        <module.author>SapphireMC</module.author>
-        <maven.deploy.skip>true</maven.deploy.skip>
-        <maven.javadoc.skip>true</maven.javadoc.skip>
-    </properties>
-
-    <dependencies>
-        <dependency>
-            <groupId>${project.groupId}</groupId>
-            <artifactId>amethyst-api</artifactId>
-            <version>${project.version}</version>
-            <scope>compile</scope>
-        </dependency>
-    </dependencies>
-
-    <build>
-        <finalName>${project.name}</finalName>
-        <resources>
-            <resource>
-                <filtering>true</filtering>
-                <directory>${basedir}/src/main/resources</directory>
-            </resource>
-        </resources>
-    </build>
-</project>
diff --git a/module/reconnect-yaml/pom.xml b/module/reconnect-yaml/pom.xml
deleted file mode 100644
index 0b36aad2..00000000
--- a/module/reconnect-yaml/pom.xml
+++ /dev/null
@@ -1,19 +0,0 @@
-
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
-    <modelVersion>4.0.0</modelVersion>
-
-    <parent>
-        <groupId>io.sapphiremc.amethyst</groupId>
-        <artifactId>amethyst-module</artifactId>
-        <version>1.18-R0.1-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
-    </parent>
-
-    <artifactId>amethyst-module-reconnect-yaml</artifactId>
-    <version>1.18-R0.1-SNAPSHOT</version>
-    <packaging>jar</packaging>
-
-    <name>reconnect_yaml</name>
-    <description>Provides reconnect location functionality in locations.yml</description>
-</project>
diff --git a/module/reconnect-yaml/src/main/java/net/md_5/bungee/module/reconnect/yaml/PluginYaml.java b/module/reconnect-yaml/src/main/java/net/md_5/bungee/module/reconnect/yaml/PluginYaml.java
deleted file mode 100644
index 0d77ce2a..00000000
--- a/module/reconnect-yaml/src/main/java/net/md_5/bungee/module/reconnect/yaml/PluginYaml.java
+++ /dev/null
@@ -1,22 +0,0 @@
-package net.md_5.bungee.module.reconnect.yaml;
-
-import net.md_5.bungee.api.config.ListenerInfo;
-import net.md_5.bungee.api.plugin.Plugin;
-
-public class PluginYaml extends Plugin
-{
-
-    @Override
-    public void onEnable()
-    {
-        // TODO: Abstract this for other reconnect modules
-        for ( ListenerInfo info : getProxy().getConfig().getListeners() )
-        {
-            if ( !info.isForceDefault() && getProxy().getReconnectHandler() == null )
-            {
-                getProxy().setReconnectHandler( new YamlReconnectHandler() );
-                break;
-            }
-        }
-    }
-}
diff --git a/module/reconnect-yaml/src/main/resources/bungee.yml b/module/reconnect-yaml/src/main/resources/bungee.yml
deleted file mode 100644
index 9757ba83..00000000
--- a/module/reconnect-yaml/src/main/resources/bungee.yml
+++ /dev/null
@@ -1,5 +0,0 @@
-name: ${project.name}
-main: net.md_5.bungee.module.reconnect.yaml.PluginYaml
-version: ${project.version}
-description: ${project.description}
-author: ${module.author}
diff --git a/pom.xml b/pom.xml
index cfbb8500..9c142f4d 100644
--- a/pom.xml
+++ b/pom.xml
@@ -45,7 +45,7 @@
         <module>event</module>
         <module>log</module>
         <module>log4j</module>
-        <module>module</module>
+       <!-- <module>module</module> Amethyst - we use new ModuleManager -->
         <module>protocol</module>
         <module>proxy</module>
         <module>query</module>
diff --git a/proxy/src/main/java/io/sapphiremc/amethyst/AmethystModulesManager.java b/proxy/src/main/java/io/sapphiremc/amethyst/AmethystModulesManager.java
new file mode 100644
index 00000000..7550d6a1
--- /dev/null
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/AmethystModulesManager.java
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2022 DenaryDev
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ */
+package io.sapphiremc.amethyst;
+
+import io.sapphiremc.amethyst.conf.AmethystConfig;
+import io.sapphiremc.amethyst.module.CommandAlert;
+import io.sapphiremc.amethyst.module.CommandAlertRaw;
+import io.sapphiremc.amethyst.module.CommandFind;
+import io.sapphiremc.amethyst.module.CommandIP;
+import io.sapphiremc.amethyst.module.CommandList;
+import io.sapphiremc.amethyst.module.CommandPerms;
+import io.sapphiremc.amethyst.module.CommandSend;
+import io.sapphiremc.amethyst.module.CommandServer;
+import io.sapphiremc.amethyst.module.YamlReconnectHandler;
+import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.api.config.ListenerInfo;
+import net.md_5.bungee.api.plugin.PluginManager;
+import net.md_5.bungee.command.CommandBungee;
+import net.md_5.bungee.command.CommandEnd;
+import net.md_5.bungee.command.CommandReload;
+
+/**
+ * Created 04.01.2022
+ *
+ * @author DenaryDev
+ */
+public class AmethystModulesManager {
+    public static void registerModules(ProxyServer proxy) {
+        AmethystConfig config = proxy.getAmethyst().getConfig();
+        PluginManager manager = proxy.getPluginManager();
+        manager.unregisterCommands(null);
+
+        manager.registerCommand(null, new CommandBungee());
+        manager.registerCommand(null, new CommandEnd());
+        manager.registerCommand(null, new CommandReload());
+
+        // Command modules
+        if (config.alertEnabled) {
+            manager.registerCommand(null, new CommandAlert());
+            manager.registerCommand(null, new CommandAlertRaw());
+        }
+        if (config.findEnabled) manager.registerCommand(null, new CommandFind());
+        if (config.ipEnabled) manager.registerCommand(null, new CommandIP());
+        if (config.listEnabled) manager.registerCommand(null, new CommandList());
+        if (config.permsEnabled) manager.registerCommand(null, new CommandPerms());
+        if (config.sendEnabled) manager.registerCommand(null, new CommandSend());
+        if (config.serverEnabled) manager.registerCommand(null, new CommandServer());
+
+        // Reconnect module
+        if (config.reconnectEnabled) {
+            for (ListenerInfo info : proxy.getConfig().getListeners()) {
+                if (!info.isForceDefault() && proxy.getReconnectHandler() == null) {
+                    proxy.setReconnectHandler(new YamlReconnectHandler());
+                    break;
+                }
+            }
+        }
+    }
+}
diff --git a/module/cmd-alert/src/main/java/net/md_5/bungee/module/cmd/alert/CommandAlert.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandAlert.java
similarity index 96%
rename from module/cmd-alert/src/main/java/net/md_5/bungee/module/cmd/alert/CommandAlert.java
rename to proxy/src/main/java/io/sapphiremc/amethyst/module/CommandAlert.java
index 423fe5bd..5f242e36 100644
--- a/module/cmd-alert/src/main/java/net/md_5/bungee/module/cmd/alert/CommandAlert.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandAlert.java
@@ -1,4 +1,4 @@
-package net.md_5.bungee.module.cmd.alert;
+package io.sapphiremc.amethyst.module;
 
 import net.md_5.bungee.api.ChatColor;
 import net.md_5.bungee.api.CommandSender;
diff --git a/module/cmd-alert/src/main/java/net/md_5/bungee/module/cmd/alert/CommandAlertRaw.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandAlertRaw.java
similarity index 97%
rename from module/cmd-alert/src/main/java/net/md_5/bungee/module/cmd/alert/CommandAlertRaw.java
rename to proxy/src/main/java/io/sapphiremc/amethyst/module/CommandAlertRaw.java
index 4b405b19..78c9a129 100644
--- a/module/cmd-alert/src/main/java/net/md_5/bungee/module/cmd/alert/CommandAlertRaw.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandAlertRaw.java
@@ -1,4 +1,4 @@
-package net.md_5.bungee.module.cmd.alert;
+package io.sapphiremc.amethyst.module;
 
 import com.google.common.base.Joiner;
 import net.md_5.bungee.api.ChatColor;
diff --git a/module/cmd-find/src/main/java/net/md_5/bungee/module/cmd/find/CommandFind.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandFind.java
similarity index 96%
rename from module/cmd-find/src/main/java/net/md_5/bungee/module/cmd/find/CommandFind.java
rename to proxy/src/main/java/io/sapphiremc/amethyst/module/CommandFind.java
index c51caf1f..bba4be04 100644
--- a/module/cmd-find/src/main/java/net/md_5/bungee/module/cmd/find/CommandFind.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandFind.java
@@ -1,4 +1,4 @@
-package net.md_5.bungee.module.cmd.find;
+package io.sapphiremc.amethyst.module;
 
 import net.md_5.bungee.api.CommandSender;
 import net.md_5.bungee.api.ProxyServer;
diff --git a/proxy/src/main/java/net/md_5/bungee/command/CommandIP.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandIP.java
similarity index 91%
rename from proxy/src/main/java/net/md_5/bungee/command/CommandIP.java
rename to proxy/src/main/java/io/sapphiremc/amethyst/module/CommandIP.java
index 9b405a0b..c6edd492 100644
--- a/proxy/src/main/java/net/md_5/bungee/command/CommandIP.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandIP.java
@@ -1,8 +1,9 @@
-package net.md_5.bungee.command;
+package io.sapphiremc.amethyst.module;
 
 import net.md_5.bungee.api.CommandSender;
 import net.md_5.bungee.api.ProxyServer;
 import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.command.PlayerCommand;
 
 public class CommandIP extends PlayerCommand
 {
diff --git a/module/cmd-list/src/main/java/net/md_5/bungee/module/cmd/list/CommandList.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandList.java
similarity index 97%
rename from module/cmd-list/src/main/java/net/md_5/bungee/module/cmd/list/CommandList.java
rename to proxy/src/main/java/io/sapphiremc/amethyst/module/CommandList.java
index 331ca78e..a7b85178 100644
--- a/module/cmd-list/src/main/java/net/md_5/bungee/module/cmd/list/CommandList.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandList.java
@@ -1,4 +1,4 @@
-package net.md_5.bungee.module.cmd.list;
+package io.sapphiremc.amethyst.module;
 
 import java.util.ArrayList;
 import java.util.Collections;
diff --git a/proxy/src/main/java/net/md_5/bungee/command/CommandPerms.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandPerms.java
similarity index 90%
rename from proxy/src/main/java/net/md_5/bungee/command/CommandPerms.java
rename to proxy/src/main/java/io/sapphiremc/amethyst/module/CommandPerms.java
index d52407c1..f7a0d4f6 100644
--- a/proxy/src/main/java/net/md_5/bungee/command/CommandPerms.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandPerms.java
@@ -1,4 +1,4 @@
-package net.md_5.bungee.command;
+package io.sapphiremc.amethyst.module;
 
 import java.util.HashSet;
 import java.util.Set;
@@ -12,7 +12,7 @@ public class CommandPerms extends Command
 
     public CommandPerms()
     {
-        super( "perms" );
+        super( "gperms", "bungeecord.command.perms" );
     }
 
     @Override
diff --git a/module/cmd-send/src/main/java/net/md_5/bungee/module/cmd/send/CommandSend.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandSend.java
similarity index 99%
rename from module/cmd-send/src/main/java/net/md_5/bungee/module/cmd/send/CommandSend.java
rename to proxy/src/main/java/io/sapphiremc/amethyst/module/CommandSend.java
index dbcfeb09..96dcd016 100644
--- a/module/cmd-send/src/main/java/net/md_5/bungee/module/cmd/send/CommandSend.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandSend.java
@@ -1,4 +1,4 @@
-package net.md_5.bungee.module.cmd.send;
+package io.sapphiremc.amethyst.module;
 
 import com.google.common.base.Joiner;
 import com.google.common.collect.ImmutableSet;
diff --git a/module/cmd-server/src/main/java/net/md_5/bungee/module/cmd/server/CommandServer.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandServer.java
similarity index 98%
rename from module/cmd-server/src/main/java/net/md_5/bungee/module/cmd/server/CommandServer.java
rename to proxy/src/main/java/io/sapphiremc/amethyst/module/CommandServer.java
index d81de590..c3116ac0 100644
--- a/module/cmd-server/src/main/java/net/md_5/bungee/module/cmd/server/CommandServer.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandServer.java
@@ -1,4 +1,4 @@
-package net.md_5.bungee.module.cmd.server;
+package io.sapphiremc.amethyst.module;
 
 import com.google.common.base.Function;
 import com.google.common.base.Predicate;
diff --git a/module/reconnect-yaml/src/main/java/net/md_5/bungee/module/reconnect/yaml/YamlReconnectHandler.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/YamlReconnectHandler.java
similarity index 98%
rename from module/reconnect-yaml/src/main/java/net/md_5/bungee/module/reconnect/yaml/YamlReconnectHandler.java
rename to proxy/src/main/java/io/sapphiremc/amethyst/module/YamlReconnectHandler.java
index 3a514a76..a703665b 100644
--- a/module/reconnect-yaml/src/main/java/net/md_5/bungee/module/reconnect/yaml/YamlReconnectHandler.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/YamlReconnectHandler.java
@@ -1,4 +1,4 @@
-package net.md_5.bungee.module.reconnect.yaml;
+package io.sapphiremc.amethyst.module;
 
 import java.io.File;
 import java.io.FileReader;
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 982f540f..3a1a16c1 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -23,6 +23,7 @@ import io.netty.channel.EventLoopGroup;
 import io.netty.util.ResourceLeakDetector;
 import io.sapphiremc.amethyst.BuildData;
 import io.sapphiremc.amethyst.Amethyst;
+import io.sapphiremc.amethyst.AmethystModulesManager;
 import io.sapphiremc.amethyst.api.protocol.ProtocolVersion;
 import java.io.File;
 import java.io.FileReader;
@@ -79,15 +80,12 @@ import net.md_5.bungee.chat.TextComponentSerializer;
 import net.md_5.bungee.chat.TranslatableComponentSerializer;
 import net.md_5.bungee.command.CommandBungee;
 import net.md_5.bungee.command.CommandEnd;
-import net.md_5.bungee.command.CommandIP;
-import net.md_5.bungee.command.CommandPerms;
 import net.md_5.bungee.command.CommandReload;
 import net.md_5.bungee.command.ConsoleCommandSender;
 import net.md_5.bungee.compress.CompressFactory;
 import net.md_5.bungee.conf.Configuration;
 import net.md_5.bungee.conf.YamlConfig;
 import net.md_5.bungee.forge.ForgeConstants;
-import net.md_5.bungee.module.ModuleManager;
 import net.md_5.bungee.netty.PipelineUtils;
 import net.md_5.bungee.protocol.DefinedPacket;
 import net.md_5.bungee.protocol.packet.PluginMessage;
@@ -173,7 +171,7 @@ public class BungeeCord extends ProxyServer
             .registerTypeAdapter( Favicon.class, Favicon.getFaviconTypeAdapter() ).create();
     @Getter
     private ConnectionThrottle connectionThrottle;
-    private final ModuleManager moduleManager = new ModuleManager();
+   // private final ModuleManager moduleManager = new ModuleManager(); Amethyst - we use our new module manager
     private final long serverStartTime; // Amethyst - add logging on startup
 
     {
@@ -232,11 +230,13 @@ public class BungeeCord extends ProxyServer
         serverStartTime = System.nanoTime(); // Amethyst - add logging on startup
 
         pluginManager = new PluginManager( this );
-        getPluginManager().registerCommand( null, new CommandReload() );
-        getPluginManager().registerCommand( null, new CommandEnd() );
-        getPluginManager().registerCommand( null, new CommandIP() );
-        getPluginManager().registerCommand( null, new CommandBungee() );
-        getPluginManager().registerCommand( null, new CommandPerms() );
+        // Amethyst start - We register some commands in our new manager
+       // getPluginManager().registerCommand( null, new CommandReload() );
+       // getPluginManager().registerCommand( null, new CommandEnd() );
+       // getPluginManager().registerCommand( null, new CommandIP() );
+       // getPluginManager().registerCommand( null, new CommandBungee() );
+       // getPluginManager().registerCommand( null, new CommandPerms() );
+        // Amethyst end
 
         if ( !Boolean.getBoolean( "net.md_5.bungee.native.disable" ) )
         {
@@ -277,9 +277,11 @@ public class BungeeCord extends ProxyServer
         bossEventLoopGroup = PipelineUtils.newEventLoopGroup( 0, new ThreadFactoryBuilder().setNameFormat( "Netty Boss IO Thread #%1$d" ).build() );
         workerEventLoopGroup = PipelineUtils.newEventLoopGroup( 0, new ThreadFactoryBuilder().setNameFormat( "Netty Worker IO Thread #%1$d" ).build() );
 
-        File moduleDirectory = new File( "modules" );
-        moduleManager.load( this, moduleDirectory );
-        pluginManager.detectPlugins( moduleDirectory );
+        // Amethyst start - we use our modules manager
+       // File moduleDirectory = new File( "modules" );
+       // moduleManager.load( this, moduleDirectory );
+       // pluginManager.detectPlugins( moduleDirectory );
+        // Amethyst end
 
         pluginsFolder.mkdir();
         pluginManager.detectPlugins( pluginsFolder );
@@ -288,6 +290,7 @@ public class BungeeCord extends ProxyServer
         config.load();
 
         Amethyst.getInstance().load(logger); // Amethyst - load our configs
+        AmethystModulesManager.registerModules(this); // Amethyst - register modules
 
         if ( config.isForgeSupport() )
         {
diff --git a/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java b/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java
index fe643a8e..431f268a 100644
--- a/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java
+++ b/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java
@@ -1,6 +1,7 @@
 package net.md_5.bungee.command;
 
 import io.sapphiremc.amethyst.Amethyst;
+import io.sapphiremc.amethyst.AmethystModulesManager;
 import net.md_5.bungee.BungeeCord;
 import net.md_5.bungee.api.ChatColor;
 import net.md_5.bungee.api.CommandSender;
@@ -26,6 +27,7 @@ public class CommandReload extends Command
         proxy.getPluginManager().callEvent( new ProxyReloadEvent( sender ) );
 
         Amethyst.getInstance().load( BungeeCord.getInstance().getLogger() ); // Amethyst - load our configs
+        AmethystModulesManager.registerModules( BungeeCord.getInstance() ); // Amethyst - register modules
 
         sender.sendMessage( ChatColor.BOLD.toString() + ChatColor.RED.toString() + "Amethyst has been reloaded."
                 + " This is NOT advisable and you will not be supported with any issues that arise! Please restart Amethyst ASAP." );
diff --git a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
index 787172e4..49836d12 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
@@ -115,23 +115,42 @@ public class YamlConfig implements ConfigurationAdapter
         }
 
         if(!doPermissions) return; // Waterfall
-        Map<String, Object> permissions = get( "permissions", null );
-        if ( permissions == null )
+
+        if ( !config.containsKey( "permissions.groups" ) )
         {
-            set( "permissions.default", Arrays.asList( new String[]
+            Map<String, Object> toConvert = get( "permissions", null );
+            List<String> converted = new ArrayList<>();
+            for ( Map.Entry<String, Object> entry : toConvert.entrySet() )
             {
-                "bungeecord.command.server", "bungeecord.command.list"
-            } ) );
-            set( "permissions.admin", Arrays.asList( new String[]
+                if ( !entry.getKey().equals( "groups" ) && !entry.getKey().equals( "users" ) && !converted.contains( entry.getKey() ) ) {
+                    set("permissions." + entry.getKey(), null);
+                    set("permissions.groups." + entry.getKey(), entry.getValue());
+                    converted.add( entry.getKey() );
+                }
+            }
+        }
+
+        Map<String, Object> permissions = get( "permissions.groups", null );
+        if ( permissions == null )
+        {
+            set( "permissions.groups.default", Arrays.asList("bungeecord.command.server", "bungeecord.command.list", "bungeecord.command.perms") ); // Amethyst - add permission to /gperms command
+            set( "permissions.groups.admin", Arrays.asList("bungeecord.command.alert", "bungeecord.command.end", "bungeecord.command.find", "bungeecord.command.ip", "bungeecord.command.reload", "bungeecord.command.send") ); // Amethyst - add more permissions
+        }
+
+        if ( config.containsKey( "groups" ) )
+        {
+            Map<String, Object> toConvert = get( "groups", null );
+            if ( toConvert != null )
             {
-                "bungeecord.command.alert", "bungeecord.command.end", "bungeecord.command.ip", "bungeecord.command.reload"
-            } ) );
+                set( "permissions.users", toConvert );
+                set( "groups", null );
+            }
         }
 
-        Map<String, Object> groups = get( "groups", null );
-        if ( groups == null )
+        Map<String, Object> users = get( "permissions.users", null );
+        if ( users == null )
         {
-            set( "groups.md_5", Collections.singletonList( "default" ) ); // Amethyst - set this to default, fix security issue
+            set( "permissions.users.md_5", Collections.singletonList( "default" ) ); // Amethyst - set this to default, fix security issue
         }
     }
 
@@ -329,10 +348,10 @@ public class YamlConfig implements ConfigurationAdapter
     public Collection<String> getGroups(String player)
     {
         // #1270: Do this to support player names with .
-        Map<String, Collection<String>> raw = get( "groups", Collections.emptyMap() );
-        Collection<String> groups = raw.get( player );
+        Map<String, Collection<String>> raw = get( "users", Collections.emptyMap() ); // Amethyst - rename it
+        Collection<String> users = raw.get( player );
 
-        Collection<String> ret = ( groups == null ) ? new HashSet<String>() : new HashSet<>( groups );
+        Collection<String> ret = ( users == null ) ? new HashSet<String>() : new HashSet<>( users );
         ret.add( "default" );
         return ret;
     }
-- 
2.34.1.windows.1

