From 2f70ff26576a150e417d916ba35abea7cdbe03cd Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Wed, 5 Jan 2022 01:24:39 +0500
Subject: [PATCH] Make min and max versions configurable


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index ff227aeb..b35221d7 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -7,6 +7,17 @@ import net.md_5.bungee.config.ConfigurationProvider;
  * Use this only to get parameter values!
  */
 public class AmethystConfig extends AbstractYamlConfig {
+    /**
+     * Minimum supported version.
+     * Default: 1.8
+     */
+    public String minVersion = "1.8";
+    /**
+     * Maximum supported version.
+     * Default: 1.18.2
+     */
+    public String maxVersion = "1.18.2";
+
     /**
      * A module that provides the /alert and /alertraw commands.
      */
@@ -49,6 +60,9 @@ public class AmethystConfig extends AbstractYamlConfig {
     @Override
     protected void loadContent()
     {
+        minVersion = getString( "versions.min", minVersion );
+        maxVersion = getString( "versions.max", maxVersion );
+
         alertEnabled = getBoolean( "modules.alert", alertEnabled );
         findEnabled = getBoolean( "modules.find", findEnabled );
         ipEnabled = getBoolean( "modules.ip", ipEnabled );
diff --git a/protocol/src/main/java/io/sapphiremc/amethyst/api/protocol/ProtocolVersion.java b/protocol/src/main/java/io/sapphiremc/amethyst/api/protocol/ProtocolVersion.java
index 720f33cb..ab87a53f 100644
--- a/protocol/src/main/java/io/sapphiremc/amethyst/api/protocol/ProtocolVersion.java
+++ b/protocol/src/main/java/io/sapphiremc/amethyst/api/protocol/ProtocolVersion.java
@@ -3,6 +3,7 @@ package io.sapphiremc.amethyst.api.protocol;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Sets;
+import io.sapphiremc.amethyst.Amethyst;
 import java.util.EnumSet;
 import java.util.HashMap;
 import java.util.List;
@@ -57,18 +58,16 @@ public enum ProtocolVersion {
     /**
      * Represents the lowest supported version.
      */
-    public static final ProtocolVersion MINIMUM_VERSION = MINECRAFT_1_8;
+    public static ProtocolVersion MINIMUM_VERSION;
     /**
      * Represents the highest supported version.
      */
-    public static final ProtocolVersion MAXIMUM_VERSION = values()[values().length - 1];
+    public static ProtocolVersion MAXIMUM_VERSION;
 
     /**
      * The user-friendly representation of the lowest and highest supported versions.
      */
-    public static final String SUPPORTED_VERSION_STRING = String
-            .format("%s-%s", MINIMUM_VERSION.getVersionIntroducedIn(),
-                    MAXIMUM_VERSION.getMostRecentSupportedVersion());
+    public static String SUPPORTED_VERSION_STRING;
 
     /**
      * A map linking the protocol version number to its {@link ProtocolVersion} representation.
@@ -90,6 +89,28 @@ public enum ProtocolVersion {
         ID_TO_PROTOCOL_CONSTANT = ImmutableMap.copyOf(versions);
     }
 
+    /**
+     * A map linking the minecraft version id to its {@link ProtocolVersion} representation.
+     */
+    public static ImmutableMap<String, ProtocolVersion> NAME_TO_PROTOCOL_CONSTANT;
+
+    public static void register() {
+        Map<String, ProtocolVersion> versions = new HashMap<>();
+        for (ProtocolVersion version : values()) {
+            for (String name : version.names) {
+                versions.putIfAbsent(name, version);
+            }
+        }
+
+        NAME_TO_PROTOCOL_CONSTANT = ImmutableMap.copyOf(versions);
+
+        MINIMUM_VERSION = getProtocolVersion(Amethyst.getInstance().getConfig().minVersion);
+        MAXIMUM_VERSION = getProtocolVersion(Amethyst.getInstance().getConfig().maxVersion);
+        SUPPORTED_VERSION_STRING = MINIMUM_VERSION.getVersionsSupportedBy().equals(MAXIMUM_VERSION.getVersionsSupportedBy()) ?
+                MINIMUM_VERSION.getVersionIntroducedIn() :
+                String.format("%s-%s", MINIMUM_VERSION.getVersionIntroducedIn(), MAXIMUM_VERSION.getMostRecentSupportedVersion());
+    }
+
     /**
      * A set containing all the protocols that the proxy actually supports, excluding special-purpose
      * "versions" like {@link #LEGACY} and {@link #UNKNOWN}.
@@ -182,6 +203,16 @@ public enum ProtocolVersion {
         return ID_TO_PROTOCOL_CONSTANT.getOrDefault(protocol, UNKNOWN);
     }
 
+    /**
+     * Gets the {@link ProtocolVersion} for the given version id.
+     *
+     * @param name the version as a string
+     * @return the protocol version
+     */
+    public static ProtocolVersion getProtocolVersion(String name) {
+        return NAME_TO_PROTOCOL_CONSTANT.getOrDefault(name, UNKNOWN);
+    }
+
     /**
      * Returns whether the protocol is supported.
      *
diff --git a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
index 7d3220f9..5e4ff7f4 100644
--- a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
+++ b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
@@ -16,11 +16,14 @@ public class WaterfallConfiguration extends Configuration {
      */
     private boolean logInitialHandlerConnections = true;
 
+    // Amethyst start - we use own methods
+    /*
     /**
      * The supported versions displayed to the client
      * <p>Default is a comma separated list of supported versions. For example 1.8.x, 1.9.x, 1.10.x</p>
      */
-    private String gameVersion;
+   // private String gameVersion;
+    // Amethyst end
 
     /**
      * Whether we use Netty's async DNS resolver for the HttpClient.
@@ -69,7 +72,7 @@ public class WaterfallConfiguration extends Configuration {
         config.load(false); // Load, but no permissions
         logInitialHandlerConnections = config.getBoolean( "log_initial_handler_connections", logInitialHandlerConnections );
         // Amethyst - Make the version look better
-        gameVersion = config.getString("game_version", "").isEmpty() ? ProtocolVersion.SUPPORTED_VERSION_STRING : config.getString("game_version", "");
+       // gameVersion = config.getString("game_version", "").isEmpty() ? ProtocolVersion.SUPPORTED_VERSION_STRING : config.getString("game_version", ""); // Amethyst - we use own methods
         useNettyDnsResolver = config.getBoolean("use_netty_dns_resolver", useNettyDnsResolver);
         // Throttling options
         tabThrottle = config.getInt("throttling.tab_complete", tabThrottle);
@@ -87,7 +90,7 @@ public class WaterfallConfiguration extends Configuration {
 
     @Override
     public String getGameVersion() {
-        return gameVersion;
+        return ProtocolVersion.SUPPORTED_VERSION_STRING; // Amethyst start - we use own methods
     }
 
     @Override
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 610911d4..b0a20273 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -291,6 +291,7 @@ public class BungeeCord extends ProxyServer
 
         Amethyst.getInstance().load(logger); // Amethyst - load our configs
         AmethystModulesManager.registerModules(this); // Amethyst - register modules
+        ProtocolVersion.register(); // Amethyst - configurable min and max versions
 
         if ( config.isForgeSupport() )
         {
diff --git a/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java b/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java
index 431f268a..9915b800 100644
--- a/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java
+++ b/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java
@@ -2,9 +2,12 @@ package net.md_5.bungee.command;
 
 import io.sapphiremc.amethyst.Amethyst;
 import io.sapphiremc.amethyst.AmethystModulesManager;
+import io.sapphiremc.amethyst.api.protocol.ProtocolVersion;
 import net.md_5.bungee.BungeeCord;
 import net.md_5.bungee.api.ChatColor;
 import net.md_5.bungee.api.CommandSender;
+import net.md_5.bungee.api.connection.PendingConnection;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
 import net.md_5.bungee.api.event.ProxyReloadEvent;
 import net.md_5.bungee.api.plugin.Command;
 
@@ -28,6 +31,23 @@ public class CommandReload extends Command
 
         Amethyst.getInstance().load( BungeeCord.getInstance().getLogger() ); // Amethyst - load our configs
         AmethystModulesManager.registerModules( BungeeCord.getInstance() ); // Amethyst - register modules
+        ProtocolVersion.register(); // Amethyst - register min and max supported versions
+
+        // Amethyst start - check players for outdated
+        if ( proxy.getOnlineCount() > 0 ) {
+            for ( ProxiedPlayer player : proxy.getPlayers() )
+            {
+                PendingConnection connection = player.getPendingConnection();
+                if ( connection.getProtocol().isAfter( ProtocolVersion.MAXIMUM_VERSION.getProtocol() ) )
+                {
+                    connection.disconnect( proxy.getTranslation( "outdated_server", proxy.getGameVersion() ) );
+                } else if ( connection.getProtocol().isBefore( ProtocolVersion.MINIMUM_VERSION.getProtocol() ) )
+                {
+                    connection.disconnect( proxy.getTranslation( "outdated_client", proxy.getGameVersion() ) );
+                }
+            }
+        }
+        // Amethyst end
 
         sender.sendMessage( ChatColor.BOLD.toString() + ChatColor.RED.toString() + "Amethyst has been reloaded."
                 + " This is NOT advisable and you will not be supported with any issues that arise! Please restart Amethyst ASAP." );
-- 
2.36.1.windows.1

