From 1d5dc99486a7afab3f47bcbf30aa9d59c15e3afb Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Mon, 10 Jan 2022 01:41:27 +0500
Subject: [PATCH] Always release packets


diff --git a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
index 4f3694db..eacc9ae0 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
@@ -116,12 +116,13 @@ public class HandlerBoss extends ChannelInboundHandlerAdapter
             return;
         }
 
-        if ( handler != null )
+        // Amethyst start - always release packets.
+        PacketWrapper packet = (PacketWrapper) msg;
+        try
         {
-            PacketWrapper packet = (PacketWrapper) msg;
-            boolean sendPacket = handler.shouldHandle( packet );
-            try
+            if ( handler != null )
             {
+                boolean sendPacket = handler.shouldHandle( packet );
                 if ( sendPacket && packet.packet != null )
                 {
                     try
@@ -136,11 +137,12 @@ public class HandlerBoss extends ChannelInboundHandlerAdapter
                 {
                     handler.handle( packet );
                 }
-            } finally
-            {
-                packet.trySingleRelease();
             }
+        } finally
+        {
+            packet.trySingleRelease();
         }
+        // Amethyst end
     }
 
     @Override
-- 
2.37.1.windows.1

