From 0e6d569e76c1592b638a19e357da6c70b5d83e6c Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Mon, 10 Jan 2022 01:51:44 +0500
Subject: [PATCH] Apply packet limits from FlameCord


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/EncryptionRequest.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/EncryptionRequest.java
index 24f94ad8..f804676d 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/EncryptionRequest.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/EncryptionRequest.java
@@ -41,4 +41,18 @@ public class EncryptionRequest extends DefinedPacket
     {
         handler.handle( this );
     }
+
+    // Amethyst start - Apply packet limits (from FlameCord)
+    @Override
+    public int expectedMaxLength(ByteBuf buf, ProtocolDirection direction, int protocolVersion)
+    {
+        return 532;
+    }
+
+    @Override
+    public int expectedMinLength(ByteBuf buf, ProtocolDirection direction, int protocolVersion)
+    {
+        return 22;
+    }
+    // Amethyst end
 }
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Handshake.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Handshake.java
index 7dbbfd3c..32c0eafd 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Handshake.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Handshake.java
@@ -1,6 +1,7 @@
 package net.md_5.bungee.protocol.packet;
 
 import io.netty.buffer.ByteBuf;
+import io.sapphiremc.amethyst.api.protocol.ProtocolDirection;
 import lombok.AllArgsConstructor;
 import lombok.Data;
 import lombok.EqualsAndHashCode;
@@ -43,4 +44,18 @@ public class Handshake extends DefinedPacket
     {
         handler.handle( this );
     }
+
+    // Amethyst start - Apply packet limits (from FlameCord)
+    @Override
+    public int expectedMaxLength(ByteBuf buf, ProtocolDirection direction, int protocolVersion)
+    {
+        return 1292;
+    }
+
+    @Override
+    public int expectedMinLength(ByteBuf buf, ProtocolDirection direction, int protocolVersion)
+    {
+        return 5;
+    }
+    // Amethyst end
 }
-- 
2.35.1

