From fbc5fcf65c6aabd27306573c43a6e59ac2749297 Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Thu, 3 Mar 2022 00:54:41 +0500
Subject: [PATCH] TCP Fast Open


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index 857b77e7..0125b4ca 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -67,6 +67,17 @@ public class AmethystConfig extends AbstractYamlConfig {
         return lobbyServers.get(random.nextInt(lobbyServers.size()));
     }
 
+    /**
+     * Enable TCP fast open netty feature.
+     * For more information see <a href="https://netty.io/wiki/tcp-fast-open.html">netty wiki</a>
+     */
+    public boolean connectionTFOEnable = true;
+
+    /**
+     * Max pending fast open
+     */
+    public int connectionTFOMaxPending = 3;
+
     /**
      * Enable or disable InitialHandler logging.
      */
@@ -136,6 +147,9 @@ public class AmethystConfig extends AbstractYamlConfig {
         lobbyMessage = getBoolean("lobby.message", lobbyMessage);
         lobbyServers = getStringList("lobby.servers", lobbyServers);
 
+        connectionTFOEnable = getBoolean("connection.tfo.enable", connectionTFOEnable);
+        connectionTFOMaxPending = getInt("connection.tfo.max-pending", connectionTFOMaxPending);
+
         logInitialHandler = getBoolean("logger.initialHandler", logInitialHandler);
         logExceptions = getBoolean("logger.exceptions", logExceptions);
         logHaProxy = getBoolean("logger.haProxy", logHaProxy);
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
index cd811cfe..45043b83 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
@@ -32,6 +32,8 @@ import java.net.SocketAddress;
 import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.TimeUnit;
 import java.util.logging.Level;
+
+import io.sapphiremc.amethyst.Amethyst;
 import net.md_5.bungee.BungeeCord;
 import net.md_5.bungee.Util;
 import net.md_5.bungee.api.ProxyServer;
@@ -193,6 +195,11 @@ public class PipelineUtils
                 // IP_TOS is not supported (Windows XP / Windows Server 2003)
             }
             ch.config().setOption( ChannelOption.TCP_NODELAY, true );
+            // Amethyst start - TFO settings
+            if ( Amethyst.getInstance().getConfig().connectionTFOEnable )
+            {
+                ch.config().setOption( ChannelOption.TCP_FASTOPEN, Amethyst.getInstance().getConfig().connectionTFOMaxPending );
+            }
             ch.config().setAllocator( PooledByteBufAllocator.DEFAULT );
             ch.config().setWriteBufferWaterMark( MARK );
 
-- 
2.37.1.windows.1

