From 0f0919a23e4d18147499e4429bee70c21d3d1833 Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Mon, 10 Jan 2022 18:39:08 +0500
Subject: [PATCH] Add lobby and hub commands


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index f534dd2c..d242d187 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -100,6 +100,28 @@ public class AmethystConfig extends AbstractYamlConfig {
      */
     public List<String> motdSecondLines = Collections.singletonList("&eChange me in &6amethyst.yml &efile!");
 
+    /**
+     * A module that provides the /lobby and /hub commands.
+     */
+    public boolean lobbyEnabled = true;
+    /**
+     * Enable or disable lobby teleport message.
+     */
+    public boolean lobbyMessage = true;
+    /**
+     * List of lobbies.
+     */
+    public List<String> lobbyServers = Collections.singletonList("lobby");
+    /**
+     * Gets random lobby server from lobby servers list.
+     */
+    public String getRandomLobby() {
+        if (lobbyServers == null) return null;
+
+        Random random = new Random();
+        return lobbyServers.get(random.nextInt(lobbyServers.size()));
+    }
+
     /**
      * Gets random motd line from first or second lines list.
      *
@@ -138,6 +160,10 @@ public class AmethystConfig extends AbstractYamlConfig {
         motdHoverBoxEnabled = getBoolean("motd.hoverBox.enabled", motdHoverBoxEnabled);
         motdHoverBoxContent = getStringList("motd.hoverBox.content", motdHoverBoxContent);
 
+        lobbyEnabled = getBoolean("lobby.enabled", lobbyEnabled);
+        lobbyMessage = getBoolean("lobby.message", lobbyMessage);
+        lobbyServers = getStringList("lobby.servers", lobbyServers);
+
         firewallEnabled = getBoolean("firewall.enabled", firewallEnabled);
         firewallNotify = getBoolean("firewall.notify", firewallNotify);
         firewallClearInterval = getInt("firewall.clear-interval", firewallClearInterval);
diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/MessagesConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/MessagesConfig.java
index 835888c3..bebfa299 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/MessagesConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/MessagesConfig.java
@@ -22,6 +22,9 @@ public class MessagesConfig extends AbstractYamlConfig {
         getString("command.list.header", "Available servers:");
         getString("command.list.entry", "&a[{0}] &e({1}): &r{2}");
         getString("command.list.total", "Total players online: {0}");
+        getString("command.lobby.success", "&7&oAttempt to move you to the lobby...");
+        getString("command.lobby.error.already-in", "&cYou are already in the lobby!");
+        getString("command.lobby.error.not-found", "&cNo lobby servers were found, please inform the server administrator!");
         getString("command.perms.groups", "&6You have the following groups: {0}");
         getString("command.perms.permissions", "=&9- {0}");
         getString("command.reload.success", "&c&lAmethyst has been reloaded. This is NOT advisable and you will not be supported with any issues that arise! Please restart Amethyst ASAP.");
diff --git a/flame/src/main/resources/amethyst.yml b/flame/src/main/resources/amethyst.yml
index ca82d370..c584a6d1 100644
--- a/flame/src/main/resources/amethyst.yml
+++ b/flame/src/main/resources/amethyst.yml
@@ -48,6 +48,19 @@ versions:
   # Default: 1.18.2
   max: "1.18.2"
 
+# Lobby module configuration
+lobby:
+  # Enable lobby module?
+  enabled: true
+
+  # Send the player a message when teleporting to the lobby?
+  message: true
+
+  # Lobby servers
+  # Use ids from bungeecord.yml
+  servers:
+    - "lobby"
+
 # Firewall configuration
 firewall:
   # Enable firewall system?
diff --git a/flame/src/main/resources/messages.yml b/flame/src/main/resources/messages.yml
index 071c698f..d4de461c 100644
--- a/flame/src/main/resources/messages.yml
+++ b/flame/src/main/resources/messages.yml
@@ -23,6 +23,13 @@ command:
     entry: "&a[{0}] &e({1}): &r{2}"
     total: "Total players online: {0}"
 
+  # Lobby command
+  lobby:
+    success: "&7&oAttempt to move you to the lobby..."
+    error:
+      already-in: "&cYou are already in the lobby!"
+      not-found: "&cNo lobby servers were found, please inform the server administrator!"
+
   # Perms command
   perms:
     groups: "&6You have the following groups: {0}"
diff --git a/proxy/src/main/java/io/sapphiremc/amethyst/AmethystModulesManager.java b/proxy/src/main/java/io/sapphiremc/amethyst/AmethystModulesManager.java
index e8c575c3..0c0bd76e 100644
--- a/proxy/src/main/java/io/sapphiremc/amethyst/AmethystModulesManager.java
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/AmethystModulesManager.java
@@ -7,6 +7,7 @@ import io.sapphiremc.amethyst.module.CommandAlertRaw;
 import io.sapphiremc.amethyst.module.CommandFind;
 import io.sapphiremc.amethyst.module.CommandIP;
 import io.sapphiremc.amethyst.module.CommandList;
+import io.sapphiremc.amethyst.module.CommandLobby;
 import io.sapphiremc.amethyst.module.CommandPerms;
 import io.sapphiremc.amethyst.module.CommandSend;
 import io.sapphiremc.amethyst.module.CommandServer;
@@ -39,6 +40,7 @@ public class AmethystModulesManager {
         if (config.findEnabled) manager.registerCommand(null, new CommandFind());
         if (config.ipEnabled) manager.registerCommand(null, new CommandIP());
         if (config.listEnabled) manager.registerCommand(null, new CommandList());
+        if (config.lobbyEnabled) manager.registerCommand(null, new CommandLobby());
         if (config.permsEnabled) manager.registerCommand(null, new CommandPerms());
         if (config.sendEnabled) manager.registerCommand(null, new CommandSend());
         if (config.serverEnabled) manager.registerCommand(null, new CommandServer());
diff --git a/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandLobby.java b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandLobby.java
new file mode 100644
index 00000000..f2da93b7
--- /dev/null
+++ b/proxy/src/main/java/io/sapphiremc/amethyst/module/CommandLobby.java
@@ -0,0 +1,38 @@
+package io.sapphiremc.amethyst.module;
+
+import io.sapphiremc.amethyst.conf.AmethystConfig;
+import java.util.List;
+import net.md_5.bungee.api.CommandSender;
+import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.api.plugin.Command;
+
+public class CommandLobby extends Command {
+
+    public CommandLobby() {
+        super("lobby", "bungeecord.command.lobby", "hub");
+    }
+
+    @Override
+    public void execute(CommandSender sender, String[] args) {
+        ProxyServer proxy = ProxyServer.getInstance();
+        if (sender instanceof ProxiedPlayer) {
+            AmethystConfig config = proxy.getAmethyst().getConfig();
+            if (!config.lobbyServers.isEmpty()) {
+                ProxiedPlayer player = (ProxiedPlayer) sender;
+                if (!config.lobbyServers.contains(player.getServer().getInfo().getName())) {
+                    player.connect(proxy.getServerInfo(config.getRandomLobby()));
+                    if (config.lobbyMessage) {
+                        player.sendMessage(proxy.getTranslationComponent("command.lobby.success"));
+                    }
+                } else {
+                    sender.sendMessage(proxy.getTranslationComponent("command.lobby.error.already-in"));
+                }
+            } else {
+                sender.sendMessage(proxy.getTranslationComponent("command.lobby.error.not-found"));
+            }
+        } else {
+            sender.sendMessage(proxy.getTranslationComponent("error.players-only"));
+        }
+    }
+}
-- 
2.35.1

