From d9e51665f88edba93881eff906e5d1a713bba79a Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Sun, 6 Mar 2022 00:39:45 +0500
Subject: [PATCH] Customizable bungee name


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index 0125b4ca..75fe48a2 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -11,6 +11,11 @@ import net.md_5.bungee.config.ConfigurationProvider;
  * Use this only to get parameter values!
  */
 public class AmethystConfig extends AbstractYamlConfig {
+    /**
+     * Custom bungee name.
+     */
+    public String bungeeName = "Amethyst";
+
     /**
      * Enable or disable custom motd.
      */
@@ -137,6 +142,8 @@ public class AmethystConfig extends AbstractYamlConfig {
     @Override
     protected void loadContent()
     {
+        bungeeName = getString("general.bungee-name", bungeeName);
+
         motdEnabled = getBoolean("motd.enabled", motdEnabled);
         motdFirstLines = getStringList("motd.firstLines", motdFirstLines);
         motdSecondLines = getStringList("motd.secondLines", motdSecondLines);
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 730332f7..c5461bd6 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -555,7 +555,7 @@ public class BungeeCord extends ProxyServer
     @Override
     public String getName()
     {
-        return "Amethyst";
+        return (Amethyst.getInstance() != null && Amethyst.getInstance().getConfig() != null) ? Amethyst.getInstance().getConfig().bungeeName : "Amethyst";
     }
 
     @Override
-- 
2.37.1.windows.1

