From 25380d5d88e6f955b3e509067d7f28179c08132a Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Thu, 3 Mar 2022 00:54:41 +0500
Subject: [PATCH] TCP Fast Open


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index 1aff5459..0f22a87e 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -78,6 +78,17 @@ public class AmethystConfig extends AbstractYamlConfig {
         return lobbyServers.get(random.nextInt(lobbyServers.size()));
     }
 
+    /**
+     * Enable TCP fast open netty feature.
+     * For more information see <a href="https://netty.io/wiki/tcp-fast-open.html">netty wiki</a>
+     */
+    public boolean connectionTFOEnable = true;
+
+    /**
+     * Max pending fast open
+     */
+    public int connectionTFOMaxPending = 3;
+
     /**
      * Enable or disable InitialHandler logging.
      */
@@ -146,6 +157,9 @@ public class AmethystConfig extends AbstractYamlConfig {
         lobbyMessage = getBoolean("lobby.message", lobbyMessage);
         lobbyServers = getStringList("lobby.servers", lobbyServers);
 
+        connectionTFOEnable = getBoolean("connection.tfo.enable", connectionTFOEnable);
+        connectionTFOMaxPending = getInt("connection.tfo.max-pending", connectionTFOMaxPending);
+
         logInitialHandler = getBoolean("logger.initialHandler", logInitialHandler);
         logExceptions = getBoolean("logger.exceptions", logExceptions);
         logHaProxy = getBoolean("logger.haProxy", logHaProxy);
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
index cd811cfe..f1b11b9d 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
@@ -193,6 +193,11 @@ public class PipelineUtils
                 // IP_TOS is not supported (Windows XP / Windows Server 2003)
             }
             ch.config().setOption( ChannelOption.TCP_NODELAY, true );
+            // Amethyst start - TFO settings
+            if ( ProxyServer.getInstance().getAmethyst().getConfig().connectionTFOEnable )
+            {
+                ch.config().setOption( ChannelOption.TCP_FASTOPEN, ProxyServer.getInstance().getAmethyst().getConfig().connectionTFOMaxPending );
+            }
             ch.config().setAllocator( PooledByteBufAllocator.DEFAULT );
             ch.config().setWriteBufferWaterMark( MARK );
 
-- 
2.36.1.windows.1

