From 48757fdf2e58314c3b7a80bb7af0de7fc2d60d86 Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Thu, 3 Mar 2022 00:54:41 +0500
Subject: [PATCH] TCP Fast Open


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index e3ba79b9..66e39af3 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -55,6 +55,17 @@ public class AmethystConfig extends AbstractYamlConfig {
      */
     public String maxVersion = "1.18.2";
 
+    /**
+     * Enable TCP fast open netty feature.
+     * For more information see https://netty.io/wiki/tcp-fast-open.html
+     */
+    public boolean connectionTFOEnable = true;
+
+    /**
+     * Max pending fast open
+     */
+    public int connectionTFOMaxPending = 3;
+
     /**
      * Enable or disable InitialHandler logging.
      */
@@ -145,6 +156,9 @@ public class AmethystConfig extends AbstractYamlConfig {
         lobbyMessage = getBoolean("lobby.message", lobbyMessage);
         lobbyServers = getStringList("lobby.servers", lobbyServers);
 
+        connectionTFOEnable = getBoolean("connection.tfo.enable", connectionTFOEnable);
+        connectionTFOMaxPending = getInt("connection.tfo.max-pending", connectionTFOMaxPending);
+
         logInitialHandler = getBoolean("logger.initialHandler", logInitialHandler);
         logExceptions = getBoolean("logger.exceptions", logExceptions);
         logHaProxy = getBoolean("logger.haProxy", logHaProxy);
diff --git a/flame/src/main/resources/amethyst.yml b/flame/src/main/resources/amethyst.yml
index 10739f05..24effcbf 100644
--- a/flame/src/main/resources/amethyst.yml
+++ b/flame/src/main/resources/amethyst.yml
@@ -61,6 +61,17 @@ lobby:
   servers:
     - "lobby"
 
+# Connection settings
+connection:
+  # TCP Fast open
+  tfo:
+    # Enable TCP fast open netty feature.
+    # For more information see https://netty.io/wiki/tcp-fast-open.html
+    enable: true
+
+    # Max pending fast open
+    max-pending: 3
+
 # Logger configuration
 logger:
   # Enable or disable InitialHandler logging.
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
index 3d08e38a..fceea075 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
@@ -197,6 +197,11 @@ public class PipelineUtils
                 // IP_TOS is not supported (Windows XP / Windows Server 2003)
             }
             ch.config().setOption( ChannelOption.TCP_NODELAY, true );
+            // Amethyst start - TFO settings
+            if ( ProxyServer.getInstance().getAmethyst().getConfig().connectionTFOEnable )
+            {
+                ch.config().setOption( ChannelOption.TCP_FASTOPEN, ProxyServer.getInstance().getAmethyst().getConfig().connectionTFOMaxPending );
+            }
             ch.config().setAllocator( PooledByteBufAllocator.DEFAULT );
             ch.config().setWriteBufferWaterMark( MARK );
 
-- 
2.35.1

