From bcf65ead1eea1b2dd32e1955d7ca47d4c851930d Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Sun, 6 Mar 2022 00:39:45 +0500
Subject: [PATCH] Customizable bungee name


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index 66e39af3..e992d378 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -136,6 +136,11 @@ public class AmethystConfig extends AbstractYamlConfig {
      */
     public List<String> motdHoverBoxContent = Arrays.asList("&eAmethyst", "", "&7Online: &f%online%", "", "&bServers:", "&8- &7Lobby: &f%online_lobby%", "", "&eexample.net");
 
+    /**
+     * Custom bungee name.
+     */
+    public String bungeeName = "Amethyst";
+
     public AmethystConfig(ConfigurationProvider provider) {
         super(provider, "amethyst.yml");
         init();
@@ -145,6 +150,8 @@ public class AmethystConfig extends AbstractYamlConfig {
     @Override
     protected void loadContent()
     {
+        bungeeName = getString("general.bungee-name", bungeeName);
+
         motdEnabled = getBoolean("motd.enabled", motdEnabled);
         motdFirstLines = getStringList("motd.firstLines", motdFirstLines);
         motdSecondLines = getStringList("motd.secondLines", motdSecondLines);
diff --git a/flame/src/main/resources/amethyst.yml b/flame/src/main/resources/amethyst.yml
index 24effcbf..dcf38c34 100644
--- a/flame/src/main/resources/amethyst.yml
+++ b/flame/src/main/resources/amethyst.yml
@@ -1,6 +1,11 @@
 # Amethyst configuration file
 # Generated on ${project.version}
 
+# General settings
+general:
+  # Custom bungee name
+  bungee-name: "Amethyst"
+
 # Motd configuration
 motd:
   # Enable amethyst custom motd?
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 9314daff..5c3f525e 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -558,7 +558,7 @@ public class BungeeCord extends ProxyServer
     @Override
     public String getName()
     {
-        return "Amethyst";
+        return (getAmethyst() != null && getAmethyst().getConfig() != null) ? getAmethyst().getConfig().bungeeName : "Amethyst";
     }
 
     @Override
-- 
2.36.0

