From 9022a3cfae5e77dc33763233918e6e37810899bd Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Sun, 6 Mar 2022 00:39:45 +0500
Subject: [PATCH] Customizable bungee name


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index c45b34f8..0cc087d0 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -11,6 +11,11 @@ import net.md_5.bungee.config.ConfigurationProvider;
  * Use this only to get parameter values!
  */
 public class AmethystConfig extends AbstractYamlConfig {
+    /**
+     * Custom bungee name.
+     */
+    public String bungeeName = "Amethyst";
+
     /**
      * Enable or disable custom motd.
      */
@@ -148,6 +153,8 @@ public class AmethystConfig extends AbstractYamlConfig {
     @Override
     protected void loadContent()
     {
+        bungeeName = getString("general.bungee-name", bungeeName);
+
         motdEnabled = getBoolean("motd.enabled", motdEnabled);
         motdFirstLines = getStringList("motd.firstLines", motdFirstLines);
         motdSecondLines = getStringList("motd.secondLines", motdSecondLines);
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 9314daff..5c3f525e 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -558,7 +558,7 @@ public class BungeeCord extends ProxyServer
     @Override
     public String getName()
     {
-        return "Amethyst";
+        return (getAmethyst() != null && getAmethyst().getConfig() != null) ? getAmethyst().getConfig().bungeeName : "Amethyst";
     }
 
     @Override
-- 
2.37.1.windows.1

