From bcd5bdca1a87d9b5eea11dac2ab168c38d5b5fe4 Mon Sep 17 00:00:00 2001
From: DenaryDev <denaryplanet@gmail.com>
Date: Sun, 6 Mar 2022 23:35:34 +0500
Subject: [PATCH] More user-friendly connection fails messages


diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
index d9db2a97..b9c48351 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/AmethystConfig.java
@@ -99,6 +99,10 @@ public class AmethystConfig extends AbstractYamlConfig {
      * Enable or disable HaProxy logging.
      */
     public boolean logHaProxy = false;
+    /**
+     * Log exception when connection to backend server failed
+     */
+    public boolean logBackendConnectionFailed = true;
 
     /**
      * A module that provides the /alert and /alertraw commands.
@@ -165,6 +169,7 @@ public class AmethystConfig extends AbstractYamlConfig {
         logInitialHandler = getBoolean("logger.initialHandler", logInitialHandler);
         logExceptions = getBoolean("logger.exceptions", logExceptions);
         logHaProxy = getBoolean("logger.haProxy", logHaProxy);
+        logBackendConnectionFailed = getBoolean("logger.backendConnectionFailed", logBackendConnectionFailed);
 
         alertEnabled = getBoolean( "modules.alert", alertEnabled );
         findEnabled = getBoolean( "modules.find", findEnabled );
diff --git a/flame/src/main/java/io/sapphiremc/amethyst/conf/MessagesConfig.java b/flame/src/main/java/io/sapphiremc/amethyst/conf/MessagesConfig.java
index d3d81ffa..641465ac 100644
--- a/flame/src/main/java/io/sapphiremc/amethyst/conf/MessagesConfig.java
+++ b/flame/src/main/java/io/sapphiremc/amethyst/conf/MessagesConfig.java
@@ -83,6 +83,8 @@ public class MessagesConfig extends AbstractYamlConfig {
         formattedMessages.put("secure_profile_expired", ChatColor.translateAlternateColorCodes('&', getString("error.secure-profile.expired", "Secure profile expired.")));
         formattedMessages.put("secure_profile_invalid", ChatColor.translateAlternateColorCodes('&', getString("error.secure-profile.invalid", "Secure profile invalid.")));
         formattedMessages.put("username_needed", ChatColor.translateAlternateColorCodes('&', getString("error.username-needed", "&cPlease follow this command by a user name.")));
+        formattedMessages.put("timeout", ChatColor.translateAlternateColorCodes('&', getString("error.timeout", "Server not reachable (timeout). Offline? Incorrectly configured address/port/firewall?")));
+        formattedMessages.put("server_unavailable", ChatColor.translateAlternateColorCodes('&', getString("error.server-unavailable", "This server is unavailable")));
     }
 
     public String getTranslation(final String name, final Object... args) {
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index ab1e3f42..d34caaa1 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -10,6 +10,7 @@ import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelInitializer;
 import io.netty.channel.ChannelOption;
+import io.netty.channel.ConnectTimeoutException;
 import io.netty.util.internal.PlatformDependent;
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
@@ -379,10 +380,10 @@ public final class UserConnection implements ProxiedPlayer
                         connect( def, null, true, ServerConnectEvent.Reason.LOBBY_FALLBACK, request.getConnectTimeout(), request.isSendFeedback() ); // Waterfall
                     } else if ( dimensionChange )
                     {
-                        disconnect( bungee.getTranslation( "fallback_kick", connectionFailMessage( future.cause() ) ) );
+                        disconnect( bungee.getTranslation( "fallback_kick", connectionFailMessage( future.cause(), target.getName() ) ) ); // Amethyst - More user-friendly connection fail messages
                     } else
                     {
-                        if (request.isSendFeedback()) sendMessage( bungee.getTranslation( "fallback_kick", connectionFailMessage( future.cause() ) ) );
+                        if (request.isSendFeedback()) sendMessage( bungee.getTranslation( "fallback_kick", connectionFailMessage( future.cause(), target.getName() ) ) ); // Amethyst - More user-friendly connection fail messages
                     }
                 }
             }
@@ -402,10 +403,25 @@ public final class UserConnection implements ProxiedPlayer
         b.connect().addListener( listener );
     }
 
-    private String connectionFailMessage(Throwable cause)
+    private String connectionFailMessage(Throwable cause, String server) // Amethyst - More user-friendly connection fail messages
     {
-        bungee.getLogger().log(Level.WARNING, "Error occurred processing connection for " + this.name + " " + Util.exception( cause, false )); // Waterfall
-        return ""; // Waterfall
+        // Amethyst start - More user-friendly connection fail messages
+        if ( bungee.getAmethyst().getConfig().logBackendConnectionFailed )
+        {
+            bungee.getLogger().log( Level.WARNING, "[" + pendingConnection.getName() + "|" + pendingConnection.getAddress() + "] Failed to connect to server " + server, cause );
+        }
+
+        if ( cause instanceof ConnectTimeoutException )
+        {
+            return bungee.getTranslation( "timeout" );
+        } else if ( cause.getClass().getName().contains( "AnnotatedConnectException" ) )
+        {
+            return bungee.getTranslation( "server_unavailable" );
+        } else
+        {
+            return cause.getClass().getName();
+        }
+        // Amethyst end
     }
 
     @Override
-- 
2.37.1.windows.1

